lib/util/cache/repository/impl/s3.ts

```diff
@@ -0,0 +1,66 @@
+import { Readable } from 'stream';
+import {
+  GetObjectCommand,
+  GetObjectCommandInput,
+  PutObjectCommand,
+  PutObjectCommandInput,
+} from '@aws-sdk/client-s3';
+import { logger } from '../../../../logger';
+import { getS3Client, parseS3Url } from '../../../s3';
+import { streamToString } from '../../../streams';
+import type { RepoCacheRecord } from '../types';
+import { RepoCacheBase } from './base';
+
+export class RepoCacheS3 extends RepoCacheBase {
+  private readonly s3Client;
+  private readonly bucket;
+
+  constructor(repository: string, url: string) {
+    super(repository);
+    this.bucket = parseS3Url(url)?.Bucket;
+    this.s3Client = getS3Client();
+  }
+
+  async read(): Promise<string | undefined> {
+    const cacheFileName = this.getCacheFileName();
+    const s3Params: GetObjectCommandInput = {
+      Bucket: this.bucket,
+      Key: cacheFileName,
+    };
+    try {
+      const { Body: res } = await this.s3Client.send(
+        new GetObjectCommand(s3Params)
+      );
+      if (res instanceof Readable) {
+        logger.debug('RepoCacheS3.read() - success');
+        return JSON.parse(await streamToString(res));
+      }
+      logger.warn(
+        `RepoCacheS3.read() - failure - expecting Readable return type got '${typeof res}' type instead`
+      );
+    } catch (err) {
+      logger.warn({ err }, 'RepoCacheS3.read() - failure');
+    }
+    return undefined;
+  }
+
+  async write(data: RepoCacheRecord): Promise<void> {
+    const cacheFileName = this.getCacheFileName();
+    const s3Params: PutObjectCommandInput = {
+      Bucket: this.bucket,
+      Key: cacheFileName,
+      Body: JSON.stringify(data),
+      ContentType: 'text/plain',
```

Comment:
shouldn't we use `application/json`?

Reply:
The app uses `'text/plain'` as well

Reply:
@Gabriel-Ladzaretti can you please check the AWS S3 documentation about this
@rarkins any special reason for that ?

Reply:
sure, change it to json as suggested, both work obviously and from what i recall, s3 can infer it by itself as well

Reply:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types#types

JSON is human readable there fore `text/plain` is also correct.
but we probably better off using the dedicated MIME type for it.

i guess this is more susceptible Browser/other services than in our case, as we do the encoding decoding ourselves.

Reply:
https://www.iana.org/assignments/media-types/application/json 😄